### PR TITLE
extract Google Analytics / Tag Manager inline JS

### PR DESCRIPTION
**Pull request by xrav3nz**
Previously I shoved all the server-rendered data into a `div`:

```html
<div class="hidden" id="id" data-server-rendered="hello world"></div>
```

This works but @SamSaffron brought up a question about HTML semantics. After looking through the docs, I decided to try out the `meta` tag In this PR:

> Metadata contains information about the page. This includes information about styles, scripts and **data to help software use and render the page**.  -- [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element#Document_metadata)

```html
<meta id="id" data-server-rendered="hello world" />
```

The only limitation with `meta` is it has to live in `head`, but it's fine 99% of the time (we can always fallback to `div`s for the other 1%).

If we are good with using the `meta` tag, I will amend my previous/future PRs.

### CODE REVIEW

**Comment by SamSaffron** [comment]
are we technically permitted to have an async attribute on these script tags?

**Comment by SamSaffron** [comment]
hmmm not point rendering anything here if it is not enabled: 

https://github.com/discourse/discourse/blob/103f9b5dc75ae4b2cb0ac985a871489d332db4b0/app/views/layouts/application.html.erb#L47-L48

**Comment by SamSaffron** [comment]
sorry error closing PR ... clicked wrong button

**Comment by SamSaffron** [comment]
I like the meta tags, I think it works nicely... thoughts @tgxworld  ?

**Comment by xrav3nz** [comment]
> are we technically permitted to have an async attribute on these script tags?

Hmm good question. A quick search didn't yield anything useful; will look into this. 

> no point rendering anything here if it is not enabled:

The guards are in the helper methods:

https://github.com/discourse/discourse/blob/85620abb710741ed83c42565b1033c492a756aeb/app/helpers/common_helper.rb#L2-L6

I could refactor this to extract the conditionals as helpers, if that's clearer i.e.

```erb
<%- if render_google_universal_analytics_code? %>
  <%= render partial: "common/google_universal_analytics"  %>
<% end %>
```

File: app/assets/javascripts/google-tag-manager.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  /* eslint-disable */
+  // prettier-ignore
+  (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
+  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
+  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
+  'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
+  })(window,document,'script','dataLayer',gtmDataElement.dataset.containerId);
+  /* eslint-enable */
+
+  window.dataLayer = [dataLayerJson];
```
Comment (tgxworld) [inline app/assets/javascripts/google-tag-manager.js]: Do we have to set the `dataLayer` attribute before loading the Google Tag Manager script as per https://developers.google.com/tag-manager/devguide#adding-data-layer-variables-to-a-page? 

screenshot from 2018-09-12 10-08-41

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -31,7 +31,7 @@ def google_universal_analytics_json(ua_domain_name = nil)
     if SiteSetting.ga_universal_auto_link_domains.present?
       result[:allowLinker] = true
     end
-    result.to_json.html_safe
```
Comment (tgxworld) [inline app/helpers/application_helper.rb]: hmm why was this removed?

Comment (xrav3nz) [reply app/helpers/application_helper.rb]: We are now rendering this json in a HTML attribute, so it has to be escaped (no longer html safe)

```html
<!-- html_safe -->
<meta data-json="{"key": "value"}">

<!-- without html_safe -->
<meta data-json="{&quot;key&quot;: &quot;value&quot;}">
```

File: app/assets/javascripts/google-tag-manager.js (hunk 1 above)
Comment (xrav3nz) [reply app/assets/javascripts/google-tag-manager.js]: Ah good catch, thanks!

**Comment by xrav3nz** [comment]
re async/preload on the script tags, I think we safe here since we are not changing how the analytic scripts are loading.

Also did a [live test](https://xrav3nz.github.io/ga-gtm-preload-test/) today:

image

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (tgxworld) [reply app/helpers/application_helper.rb]: :+1:

**Comment by SamSaffron** [comment]
Cool, merging this in, I want to see some of this in action 👍
